osh/hub/scripts/osh-retention

```diff
@@ -0,0 +1,99 @@
+import datetime
+import logging
+import os
+import shutil
+from datetime import date
+
+os.environ['DJANGO_SETTINGS_MODULE'] = 'osh.hub.settings'
+
+
+def deleteScansResults(scansToDelete):
+    from kobo.hub import models
+    for scan in scansToDelete:
+        try:
+            pathToDelete = models.Task.get_task_dir(scan.id)
+            if os.path.isdir(pathToDelete):
+                shutil.rmtree(pathToDelete)
+        except OSError as e:
+            logging.error(f"Error deleting scans: {e}")
+
+
+# In order to identify personal scans, we are using the owner field. If the owner doesn't contain any / or @, we assume
+# that is a personal scan
+def identifyPersonalScan(owner):
+    # We could use regexps here if we want to make this more complex but for this specific case
+    # built-in string methods are enough and we are more perfomance-oriented.
+    if "/" in owner or "@" in owner:
+        return False
+    else:
+        return True
+
+
+def main():
+    import django
+    django.setup()
+
+    # Retention policy Scans
+    print("Starting Retention Policy Enforcement")
+
+    from osh.hub.scan import models
+    tasks = models.Task.objects.all()
+
+    # ---------- Section to take parameters from DB
+
+    retentionPersonalScans = 14
+    retentionFailedStatus = 14
+
+    if models.RetentionPolicySetting.objects.filter(category="PersonalScan").exists():
+        retentionPersonalObject = models.RetentionPolicySetting.objects.get(category="PersonalScan")
+        retentionPersonalScans = retentionPersonalObject.days
+    else:
+        # We set up the default time
+        retentionPersonalScans = 14
+        logging.warning("Error getting PersonalScan parameter. Setting default parameter")
+
+    if models.RetentionPolicySetting.objects.filter(category="FailedStatus").exists():
+        retentionFailedObject = models.RetentionPolicySetting.objects.get(category="FailedStatus")
+        retentionFailedStatus = retentionFailedObject.days
+    else:
+        # We set up the default time
+        retentionFailedStatus = 14
+        logging.warning("Error getting FailedStatus parameter. Setting default parameter")
+
+    # ---------- End Section to take parameters from DB
+
+    eligiblePersonalScans = []
+    eligibleFailedStatus = []
+
+    # We identify which scans are eligible to be deleted
+    today = date.today()
+    for t in tasks:
+        # At the moment, we only have 2 rule for retention settings:
+        # Personal Scans
+        # Failed status
+        calculatedDatePersonalScans = (t.dt_created + datetime.timedelta(days=retentionPersonalScans)).date()
+        calculatedDateStatusFailed = (t.dt_created + datetime.timedelta(days=retentionFailedStatus)).date()
+        if calculatedDatePersonalScans <= today:
+            if identifyPersonalScan(t.owner.username):
+                # Personal Scans eligible to be deleted
+                eligiblePersonalScans.append(t)
+        elif calculatedDateStatusFailed <= today:
```

Comment:
I believe we want to remove non-personal tasks that that are older than `calculatedDatePersonalScans`, right?  This is not the case with the current code, though.  To fix it, I'd use a following code structure:

```python
if <task is personal>:
    <check and mark personal task for removal>
    continue

<check and mark non-personal task for removal>
```

Reply:
I don't really understand your comment here. Nevertheless, I made the code clearer:

if <task is personal>:
    <check and mark personal task for removal>
if <task is failed>:
<check and mark failed task for removal>

In previous comment, we checked before if it could be deleted because of the time constaints and then we would check the category.

Reply:
Thank you!  Your recent changes actually fixed the bug I was referring to!  Just for the sake of completeness:

```python
if calculatedDatePersonalScans <= today:
    if identifyPersonalScan(t.owner.username):
        ...
elif calculatedDateStatusFailed <= today:
        ...
```
Due to the `elif` a failed non-personal task that satisfied the personal task limit would not be deleted even when it also satisfies the failed limit.